tensorflow/lite/micro/examples/mnist_lstm/train.py

```diff
@@ -0,0 +1,166 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =============================================================================
+# pylint: disable=line-too-long
+"""
+LSTM model training for MNIST recognition
+"""
+import argparse
+import os
+import numpy as np
+import tensorflow as tf
+
+
+def create_model(units=20):
+  """Create a keras LSTM model for MNIST recognition
+
+    Args:
+        units (int, optional): dimensionality of the output space for the model. Defaults to 20.
+
+    Returns:
+        tf.keras.Model: A Keras LSTM model
+    """
+
+  model = tf.keras.models.Sequential([
+      tf.keras.layers.Input(shape=(28, 28), name='input'),
+      tf.keras.layers.LSTM(units, return_sequences=True),
+      tf.keras.layers.Flatten(),
+      tf.keras.layers.Dense(10, activation=tf.nn.softmax, name='output')
+  ])
+  model.compile(optimizer='adam',
+                loss='sparse_categorical_crossentropy',
+                metrics=['accuracy'])
+  model.summary()
+  return model
+
+
+def get_train_data():
+  """Get MNIST train and test data
+
+    Returns:
+        list of tuples: (data, label) pairs for train and test
+    """
+  (x_train, y_train), _ = tf.keras.datasets.mnist.load_data()
+  x_train = x_train / 255.  # normalize pixel values to 0-1
+  x_train = x_train.astype(np.float32)
+  return (x_train, y_train)
+
+
+def train_lstm_model(epochs):
+  """Train keras LSTM model on MNIST dataset
+
+    Args:
+        epochs (int) : number of epochs to train the model
+
+    Returns:
+        tf.keras.Model: A trained keras LSTM model
+    """
+  model = create_model()
+  x_train, y_train = get_train_data()
+  callback = tf.keras.callbacks.EarlyStopping(
+      monitor='val_loss',
+      patience=1)  #early stop if validation loss does not drop anymore
+  model.fit(x_train,
+            y_train,
+            epochs=epochs,
+            validation_split=0.2,
+            batch_size=32,
+            callbacks=[callback])
+  return model
+
+
+def save_tf_model(model, save_path):
+  """Save the trained LSTM model in tensorflow format
+
+    Args:
+        model (tf.keras.Model): the trained LSTM Model
+        save_path (str): path to save the model
+    """
+
+  batch_size, steps, input_size = 1, 28, 28
+  run_model = tf.function(lambda x: model(x))
+  concrete_func = run_model.get_concrete_function(
+      tf.TensorSpec([batch_size, steps, input_size], model.inputs[0].dtype))
+  model.save(save_path, save_format="tf", signatures=concrete_func)
```

Comment:
Why do we need to use concrete function here? If `model.save(save_path)` is equivalent I'd prefer to keep it simple.

Reply:
For some reason we need to fixed the input size for the LSTM model to make the conversion work. See https://colab.sandbox.google.com/github/tensorflow/tensorflow/blob/master/tensorflow/lite/examples/experimental_new_converter/Keras_LSTM_fusion_Codelab.ipynb#scrollTo=tB1NZBUHDogR

For the use of concrete function at the export phase, see: 
https://www.tensorflow.org/guide/saved_model#specifying_signatures_during_export 

FYI: What is a concrete function https://www.tensorflow.org/guide/function#tracing

Reply:
? I believe this path is not exercised by the converter as is now (unlike in the colab), so I'm not sure what you mean by making the conversion work. `train_lstm_model()` returns a keras model which is fed directly to the converter without going through this path. 

Just checked out your branch, seems like `model.save(save_path)` works (model is saved without error) here. Let me know if I'm missing something else.

Reply:
yes `model.save(save_path)` works without any problem. But if people want to do conversion from the saved tf model using `converter = tf.lite.TFLiteConverter.from_saved_model(MODEL_DIR)` then `tflite_model = converter.convert()` won't work (e.g., debugging or learning). My understanding is that the reason we want to have the tf model saved is to comparing it to the converted tflite model. I just thought it might be good to make sure the saved tf model can be converted, and has the same data flow as the tflite model. 

Alternative, we can follow b/225231544 instead. In this way we dont need to wrap the model to make  `tflite_model = converter.convert()` work. The problem is that I dont understand the setting

Reply:
Okay I see. If that's an intended feature, I think we should keep the keras and conversion paths the same to keep the diff minimal, i.e. use the same workaround at the beginning of save_tflite_model, so something like:
```
trained_model.save(save_path)
reconstructed_model = tf.keras.models.load_model(save_path)
save_tflite_model(reconstructed_model, save_path, False) 
```
I think it's also fine to stick with the colab example with concrete functions! It's better to keep the keras -> tflite and keras -> SavedModel -> tflite paths the same. Sorry, I didn't realize yesterday there was an issue with input fixing when I suggested using `from_keras_model`. It's up to you.

Reply:
I am confused on this.

`trained_model.save(save_path)
reconstructed_model = tf.keras.models.load_model(save_path)
save_tflite_model(reconstructed_model, save_path, False) `

This still saves the intermediate model. Do you mean we always get the tf model saved? Here, input fixing before `from_keras_model` is the equivalent to `get_concrete_function` with size fixed I believe (I might be wrong).